### PR TITLE
c&r : show alert msg if crop module is enabled

### LINKED ISSUE #9846: FR: Show an alert message if "crop and rotate" module is focused and "crop" ist already switched on
Started as a bug request, this issue was changed to a feature request...

**Did you buy darktable from an application store ?**  -> NO

**Describe the bug/issue**

**To Reproduce**
_Please provide detailed steps to reproduce the behaviour, for example:_
1. Open picture in Darkroom
grafik

2. Crop picture by "crop" module
grafik

3. Click on module "crop and rotate"
4. The complete picture is shown, not the cropped one
grafik

If you crop it again, it results in a double cropped picture, but I couldn't find the system how the second crop is done on the picture.

**Expected behavior**
If crop and rotate is activated, the cropped pictures should be shown. The other way round (first crop&rotate then crop) is working as expected.

**Platform**
_Please fill as much information as possible in the list given below. Please state "unknown" where you do not know the answer and remove any sections that are not applicable _

* darktable version : 3.7.0+684~g40421d57d
* OS : Windows 10 // 19042.1165
* Memory : 16 GB
* Graphics card : Geforce RTX2070 + Intel UHD Graphics 630
* Graphics driver : 430.81
* OpenCL installed : yes
* OpenCL activated : yes

**Additional context**
_Please provide any additional information you think may be useful, for example:_

 - Can you reproduce with another darktable version(s)? **yes with version 3.6**
 - Are the steps above reproducible with a fresh edit (i.e. after discarding history)? **yes**

### PR DESCRIPTION
**Pull request by AlicVB**
this fix #9846 

As usual, feel free to propose a better alert text if needed...

### CODE REVIEW

**Comment by domosbg** [comment]
@AlicVB Thanks a lot for your fix. I tried it and in general it works. But the alert message only appears if the crop & rotate module is first switched on (by clicking on the power button) and then focused by clicking on the headline.

I think it would be better, if the message also appears if the c&r module is only focused without switching it on before (seems to be I'm the only one who focuses the modules without switching them on before. Sorry for my unusual workflow :-) )
That's the reason why I put the alert message in my code snippet in an separate if(in) in front of the if(self->enabled).

**Comment by AlicVB** [comment]
@domosbg : you're right. I've updated this PR with you're proposal... Thanks for your tests and your code... What is here is nothing more than a copy-paste of your work ;)

**Comment by domosbg** [comment]
@AlicVB looks good for me now. Thanks again for your work.
I'm happy if my code snippet helped you to implement it or at least saved a little bit of your time.

File: src/iop/clipping.c [hunk 1]
```diff
@@ -1348,6 +1348,24 @@ void gui_focus(struct dt_iop_module_t *self, gboolean in)
...
+  if(in)
+  {
+    for(GList *l = g_list_last(self->dev->history); l; l = g_list_previous(l))
+    {
+      dt_dev_history_item_t *item = (dt_dev_history_item_t *)l->data;
+      if(!g_strcmp0(item->op_name, "crop"))
+      {
+        if(item->enabled)
+          dt_control_log(
+              _("cropping image in both 'crop' and 'crop &amp; rotate' modules may produce unexpected results"));
```
Comment (TurboGit) [inline src/iop/clipping.c:1363]: In such case why not use the alert message circuitry already available? I'm talking about what we do for color calibration and white balance?

Comment (AlicVB) [reply src/iop/clipping.c]: you're right. PR updated to use in-module trouble message...

**Comment by TurboGit** [comment]
I'm still a bit worried about this.
1. the error message is asymmetric (no message in crop module)
2. if I used only the rotation in c&r and no cropping I still get the message (is that expected?)

The last point is important as one can use c&r for rotate or perspective correction and crop late with crop module, or is that still a cause of issues?

**Comment by johnny-bit** [comment]
The correct use is do crop&rotate first (since it's early in the pipe) and then final crop (later in the pipe). So IMO the no message in crop module is correct. mesage for crop&rotate module is correct even in case of just applying rotate since it will change roi for crop module changing crop module crop value cropping weirdly crop wanted by crop making user.

**Comment by TurboGit** [comment]
@johnny-bit : Thanks for clarifying, indeed I was confused by the goal here. Looks good then.

**Comment by AlicVB** [comment]
and more important : the biggest problem when cropping in c&r with crop enabled is that what will be shown on screen doesn't correspond to the crop which will be applied at the end. hence the risk of "unexpected" results.
That's not the case with the crop module which comes later in the pipe.

Anyway, I really hope that for the next release we will be able to deprecate c&r. I "only" need to port the "drawn" perspective and we should be ok !

Thanks for your tests

**Comment by TurboGit** [comment]
My testing was not good enough it seems, now I have the alert message activated when I do something else. I had enabled c&r and did some straitening. Then I have cropped with crop module. Later using the global color picker I had the alert message starting to display. But it seems to happen in many cases...

**Comment by AlicVB** [comment]
hmmm... where do you see the alert message ?
I guess it's in the terminal...
I've copy-paste what is done in temperature.c but maybe it's a little bit too verbose...  can you see if commenting line 2143 in clipping.c is better ? (we will miss the case where crop is activated without loosing focus in c&r but that's a corner case !)

**Comment by domosbg** [comment]
Sorry that I caused so much work. Thought it would be easier to implement it. The first implementation was not so complicated, but the new version with the message on the right in the module looks much better then the popup in the picture. And if it is the prefered way to show module errors, it should be used.

Because I caused the whole work, i will try to help fixing as good as I can. Again: I'm not very familar with C. So if it makes no sense, please ignore.

Actually there are messages in the terminal also when there is switched away from the crop & rotate module . To avoid this I check if c&r is expanded:
grafik

Besides
_display_crop_error I only call in gui_update (line 2070)
I commented it out in gui focus (line 1391), in gui_changed (line 1903) and in _develop_ui_pipe_finished_callback (line 2146).

I tested this and IMO it seems to work. The red alert message is also shown if c&r is expanded if you switch from the lighttable to darkroom. Downsides I found is that in terminal the error message is shown twice if you switch from the lighttable to the darkroom and one terminal message is triggered if you switch to another module group page. Besides your corner case doesn't work. In all other cases I tested one message is triggered as expected.

To solve the corner case, I changed the code an only called _display_crop_error in  _develop_ui_pipe_finished_callback. That helps for the corner case, but it doesn't work when switching from lighttable to darkroom and the alert message is only shown after the pipe is processed, that takes a little time.

So I think the first solution would be better. Maybe that helps you to fix.

Thanks a lot for your work.

**Comment by TurboGit** [comment]
Yes, I was thinking about this and indeed we don't want to display the message if c&r is not expanded.

```
if(trouble && self->expanded)
{
}
```

Or most probably only if it has focus.

**Comment by AlicVB** [comment]
I agree with you here, but there's a quite problematic side issue I've just discovered (which is way more annoying imho) : 
When the trouble message is shown, try changing the rotation angle (for example) and see how the ui is "jumping". I've even manage to make the trouble message disappear ! this make working with the module quite difficult (double-clicking to reset the rotation is almost impossible for example)

I've filled a bug/FR about this in #9921 For our case, I would say either we implement something like what I propose in the issue, either we may have to switch back to the log message here...

**Comment by domosbg** [comment]
I can confirm it. Strange behaviour. But it is only if you make a rotation. Do you know the reason why it happens if you rotate and not if you crop? Is the pipe only new calculated if a rotation is made and then the trouble message from the other modules deletes the message from c&r?

**Comment by AlicVB** [comment]
if you make a rotation, the image on screen is changed. That means the pipe is recomputed... and the problem with the trouble message appears.
That's not the case for cropping, as what is changed is just the overlays on top of the image.

afaict, there's no way for other modules to interfere with the trouble message. The issue here is that we have 2 different trouble messages for the same module !

**Comment by domosbg** [comment]
Ok, then I missunderstood something.

Why do you write that we have 2 different trouble messages for the same module? There is only the one message in the module in _display_crop_error as far as I can see. This trouble message is set or reset in the function.

For me it is not clear why the message disappears sometimes. The message should only disappear if dt_iop_set_module_trouble_message(self, NULL, NULL, NULL) is called. And that didn't happened in my testing if I can trust the debugger.

**Comment by domosbg** [comment]
I tried something else.

I set breakpoints in the if(trouble) and the corresponding else part. There is always called the if(trouble) part. But sometimes the message is shown and sometimes not. Besides sometimes the message disappears without stopping on one of the breakpoints.
Can you tell me the reason for that?

**Comment by AlicVB** [comment]
the other "trouble" message/removal comes from : https://github.com/darktable-org/darktable/blob/b631eac0af12a08825d43d185eed2afc9a32e5ee/src/iop/clipping.c#L990
which ends to : https://github.com/darktable-org/darktable/blob/b631eac0af12a08825d43d185eed2afc9a32e5ee/src/develop/imageop.c#L3145 (in the case where the format is ok, which is the case for the majority of cases)

As you can see, the module structure is passed to the trouble routines each time...

**Comment by domosbg** [comment]
Thanks a lot for your time. Now I understand it. 

I think your suggestion in #9921 with the trouble id is a good idea and would solve the problem we encountered here.

**Pull request merged**